ui/src/warp_runner.rs

```diff
@@ -0,0 +1,198 @@
+use std::{
+    ops::{Deref, DerefMut},
+    path::PathBuf,
+    sync::Arc,
+};
+
+use tokio::sync::{
+    mpsc::{UnboundedReceiver, UnboundedSender},
+    Mutex, Notify,
+};
+use warp::{
+    constellation::Constellation, multipass::MultiPass, raygun::RayGun, tesseract::Tesseract,
+};
+use warp_fs_ipfs::config::FsIpfsConfig;
+use warp_mp_ipfs::config::MpIpfsConfig;
+use warp_rg_ipfs::{config::RgIpfsConfig, Persistent};
+
+use crate::DEFAULT_PATH;
+
+pub type WarpCmdTx = UnboundedSender<WarpCmd>;
+pub type WarpCmdRx = Arc<Mutex<UnboundedReceiver<WarpCmd>>>;
+pub type WarpEventTx = UnboundedSender<WarpEvent>;
+pub type WarpEventRx = Arc<Mutex<UnboundedReceiver<WarpEvent>>>;
+
+pub enum WarpEvent {
+    None,
+}
+
+pub enum WarpCmd {
+    None,
+}
+
+pub struct WarpRunner {
+    notify: Arc<Notify>,
+    ran_once: bool,
+}
+
+impl std::ops::Drop for WarpRunner {
+    fn drop(&mut self) {
+        self.notify.notify_waiters();
+    }
+}
+
+impl WarpRunner {
+    pub fn init() -> Self {
+        Self {
+            notify: Arc::new(Notify::new()),
+            ran_once: false,
+        }
+    }
+
+    // spawns a thread which will terminate when WarpRunner is dropped
+    pub fn run(&mut self, tx: WarpEventTx, rx: WarpCmdRx) {
+        if self.ran_once {
+            panic!("WarpRunner called run() multiple times");
```

Comment:
Could use an assert here instead. Eg `assert!(self.run_once, "WarpRunner called run() multiple times");`

Reply:
thanks! i feared assert! would be compiled out but turns out that isn't true.